### PR TITLE
 Add drop_remainder argument to bucket_by_sequence_length

### PR DESCRIPTION
**Pull request by yweweler**
`tf.data.experimental.bucket_by_sequence_length` does not allow to drop
the last batch in case it has fewer than `batch_size` elements.
This patch does implement `drop_remainder` for `bucket_by_sequence_length`
to enable thhis behaviour.

`drop_remainder` is optinal and set to `False` by default to maintain
compatibility.

I will add a PR to update the documentation of bucket_by_sequence_length once this is merged.

### CODE REVIEW

**Review by mrry — CHANGES_REQUESTED** [review]
Can you please add a test for the new arguments?

**Comment by yweweler** [comment]
@mrry , I added `drop_remainder` tests to the existing test cases.
Are these sufficient?

I run the tests using the nightly-docker file (CPU only).

File: tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py [hunk 1]
```diff
@@ -306,16 +422,18 @@ def _compute_batches(dataset):
...
+      dataset = _build_dataset()
+      boundaries = range(min_len + bucket_size + 1, max_len, bucket_size)
+      dataset = dataset.apply(grouping.bucket_by_sequence_length(
+          _element_length_fn,
+          boundaries,
+          [batch_size] * (len(boundaries) + 1),
+          no_padding=True,
+          drop_remainder=drop_remainder))
+      batches = _compute_batches(dataset)
+      expected_batches = _compute_expected_batches()
```
Comment (mrry) [inline tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py]: Should this call take a `drop_remainder` argument?

Comment (yweweler) [reply tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py]: Yap, sorry about that, it slipped into the pylint commit 6842678

**Review by mrry — DISMISSED** [review]

File: tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py [hunk 2]
```diff
@@ -154,9 +154,9 @@ def _test_bucket_by_padding(no_padding):
                                                      batch_sizes,
                                                      n_bucket_elements):
         # Initialize the sum across all batches.
-        expected_sums[length] = 0
+        generated_sums[length] = 0
         # Initialize the individual batch sizes.
-        expected_batch_sizes[length] = []
+        generated_batch_sizes[length] = []
```
Comment (yweweler) [inline tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py]: I initialized the wrong variables due to a copy paste failure.

**Comment by yweweler** [comment]
@mrry  The `bucket_by_sequence_length` test passes with the latest commit 4ddd2ab.

The only other test that is failing is the `api_compatibility_test`.
However, I am not sure how these work. Am I responsible for taking action when they fail?

**Comment by mrry** [comment]
@yweweler I think you can update the necessary API files by running the following two commands:

```
bazel build //tensorflow/tools/api/tests:api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens=True
```

**Comment by yweweler** [comment]
> @yweweler I think you can update the necessary API files by running the following two commands:
> 
> ```
> bazel build //tensorflow/tools/api/tests:api_compatibility_test
> bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens=True
> ```

@mrry I have to admit that I am a bit stuck.
I am using the two lines from [tensorflow/tools/api/tests/README.txt](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/tests/README.txt) to build and update the goldens.
```
$ bazel build //tensorflow/tools/api/tests:api_compatibility_test
$ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
```
The build part works flawlessly.
I assured that all dependencies are installed properly and that building tensorflow as a whole also works.
However, updating the goldens using `api_compatibility_test` crashes.

```
(tensorflow_27_venv) user:~/downloads/tensorflow$ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True
Traceback (most recent call last):
  File "/home/user/downloads/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 36, in <module>
    from tensorflow._api.v2 import v2 as tf_v2
  File "/home/user/downloads/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/_api/v2/v2.py", line 325, in <module>
    if _tf_api_dir not in __path__:
NameError: name '__path__' is not defined
```

There are some issues in the repo suggesting that the procedure can only be executed using python2.7.
However, I tried doing the whole procedure both using python2.7 and python3.5 and ran into the same error every time.

I also tried the same thing on older commits from the `master` branch. Sadly with the same outcome.

**Comment by yweweler** [comment]
> NameError: name '\_\_path__' is not defined

I already fiddled around with the "\_\_path__"  variable by manually initializing it inside `/home/user/downloads/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/_api/v2/__init__.py`

However, this just leads to more errors, like:
```
ImportError: cannot import name v2
```

**Comment by mrry** [comment]
@annarev Is thie `api_compatibility_test` currently expected to work in open source?

**Comment by yweweler** [comment]
> @annarev Is thie `api_compatibility_test` currently expected to work in open source?

I would like to prevent this to fall into decline.

Guessing from the `api_compatibility_test` code I think that the changes to the goldens are minimal.
Therefore I could patch the goldens manually and see if the test passes. I don't really like working around this but currently I see no other option.

Either way it would be great to have some clarification on the state of `api_compatibility_test`.

**Comment by yweweler** [comment]
@mrry I patched the goldens for both v1 and v2 (Not sure if this is desired).

I also retried executing the test with different docker images (`devel-py3` and `nightly-devel`), sadly with the  same outcome as described above.

As I can't test the changes using `api_compatibility_test`, it would be great if we could run the through kokoro.

**Comment by harshini-gadige** [comment]
@jsimsa -  Could you PTAL at this ?

File: tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py [hunk 3]
```diff
@@ -71,6 +71,130 @@ def _get_record_shape(sparse):
...
+                                 sorted(generated_batch_sizes[l])))
+
+      # Make sure the generated sequence lengths appear as often as expected.
+      self.assertEqual(sorted(generated_lengths), sorted(expected_lengths),
+                       'The generated sequence lengths did not match! '
+                       'expected: {}, generated: {}'
+                       .format(sorted(expected_lengths),
+                               sorted(generated_lengths)))
+
+    for no_padding in (True, False):
```
Comment (jsimsa) [inline tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py]: use `@parameterized` for this (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/kernel_tests/map_and_batch_test.py for an example)

**Review by jsimsa — CHANGES_REQUESTED** [review]
did a pass and left some minor comments

**Comment by yweweler** [comment]
@jsimsa I made the changes you requested.

**Comment by jsimsa** [comment]
@shivaniag could you please review this PR, the newly introduced test still uses session and I believe it should be switched to the new style which you introduced

**Comment by yweweler** [comment]
> @shivaniag could you please review this PR, the newly introduced test still uses session and I believe it should be switched to the new style which you introduced

Thats a good point. I did not notice that @shivaniag updated the tests in the meantime.

I was going through his commit 4caa17185fb39420c87373af77b538e9016be46d trying to understand the changes he made to the other tests.
Before updating the PR to match his changes I need to know why the following change is needed.

https://github.com/tensorflow/tensorflow/blob/4caa17185fb39420c87373af77b538e9016be46d/tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py#L128-L130

@shivaniag Why do we need to disable the sum_check when not executing eagerly?

**Comment by shivaniag** [comment]
@yweweler: No, we don't need to disable sum_check when executing eagerly, there is a need to add eager coverage for this test which has been left as a TODO in the last commit.

File: tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py [hunk 4]
```diff
@@ -72,10 +74,141 @@ def _get_record_shape(sparse):
...
 
 @test_util.run_all_in_graph_and_eager_modes
-class BucketBySequenceLengthTest(test_base.DatasetTestBase):
+class BucketBySequenceLengthTest(test_base.DatasetTestBase, parameterized.TestCase):
+
+  @parameterized.named_parameters(
+    ("WithoutPadding", True),
+    ("WithPadding", False),
+  )
+  def testBucketDropReminder(self, param_no_padding):
```
Comment (shivaniag) [inline tensorflow/python/data/experimental/kernel_tests/bucket_by_sequence_length_test.py]: make this testSkipEagerBucketDropReminder for now and add a TODO note for eager coverage.

**Comment by yweweler** [comment]
@shivaniag  This should implement the changes the requested changes.

**Review by shivaniag — DISMISSED** [review]
Approving the test changes.

**Comment by yweweler** [comment]
Not sure why I can't see the test results from the kokoro run any more.
However, apart from me having forgot to run pylint on the last commit, the other errors were not related to this PR (as far as I can tell.)

**Review by jsimsa — APPROVED** [review]

**Pull request merged**